### ISSUE TITLE
Use golang as part of IOS and android clients

### ISSUE THREAD

**Issue by gedw99** [centrifugal/centrifuge-android#2]
gomobile allows to easily compile for IOS and Android.

The low hanging fruit is to write the proxy client code in golang and cross compile it to IOS and android.
Its not hard these days.

This is a suggestion, because i have done this on other projects, but i dont use centrifugal, and dont have time to do it.

**Comment by SammyVimes** [comment]
What is the motivation behind this? Golang code will be compiled to a shared library and for android it's necessary to have a `so` for every processor architecture.

**Comment by gedw99** [comment]
Motivations:
- Less code
- strong typing from the bottom higher up the stack, and so DRY
- easier to test, because a big chunk of the code can be tested outside the mobile execution environment.

Your assertion about .so and needing different ISA chipset architecture targeting is true. 
I cant remember off the top of my head, but it can generate the ARM and x86 for you in one hit i think. 
You simple compile the golang code, and link to the Android and IOS GUI code.
The gomobile command will generate the bindings between the golang code and the GUI code automatically.
You can also take the approach of using a socket between the GUI and the golang code. But i doubt thats suitable here.

For any GUI developer, its very very easy. 

There is not that much doc here, but there are lots of example showing the things i said above n the code base.
https://github.com/golang/mobile

**Comment by FZambia** [comment]
@gedw99 thanks for pointing at this - this is a very interesting area to investigate btw. I thought about gomobile before but was frightened by pure exported type limits so never had a chance to make proof of concept how it suits for Centrifugo client. Specifically I remember that it had no support for type aliases. Maybe we can go without them - need to investigate. And I am not sure how it works with pointer types at moment?

But current iOS and Android clients designed in a way that is convenient for iOS and Android developers - so I suppose they both are easier to dive in and start using.

Maybe this is issue should be addressed on https://github.com/centrifugal/centrifuge-go first - i.e. try to make go shared object with Centrifugo client core.

**Comment by gedw99** [comment]
Yes pointers and some types are a problem. Very true.
The gomobile tram are working on that aspect.
But it's also possible to work around them with small adjustments too.
As always depends if what gomobile offers is worth it for you.
I think it is because you have only one code base ;)

Ios and android programmers can very easily link to you code btw. There are
quite a few examples with integration seamlessly into xcode and android
studio / grade.

You are correct in that the other way to do this is to use websockets. I
have done it this way also.
Basically you putting the server proxy on the mobile, and the third party
android / ios or react native or web site ( using gopherjs) just talks to
it over start date WS.
It works and gets huge "DRYness" :)

When I build stuff I also just use gob serialisation itself with pure
golang rpc. It's core of golang so nice and easy.
The clients can be given a proxy that speaks this, and use it via FFI.
Another way to consider. It reduced the code even more.

**Comment by FZambia** [comment]
So generally I like an idea to implement Centrifugo client core in gomobile and will try to monitor its development to come up with proof of concept in future.

Closing because it's not related to centrifuge-android in its current design.
